File: Synapse/src/lib/resume-parser.ts
import type { ResumeData } from '@/types';

/**
 * Extracts plain text from a File (PDF or DOCX).
 * For PDF/DOCX we extract raw text via browser APIs; for .txt we read directly.
 * Full PDF/DOCX parsing in-browser requires heavy libs; we use a pragmatic approach:
 * read as text and let the pattern parser extract structure.
 */
export async function extractTextFromFile(file: File): Promise<string> {
  if (file.type === 'text/plain' || file.name.endsWith('.txt')) {
    return await file.text();
  }

  // For PDF/DOCX, attempt to read as text (works for simple files).
  // In production this would use pdfplumber/python-docx server-side.
  try {
    const text = await file.text();
    if (text && text.trim().length > 50) {
      return text.replace(/[^\x20-\x7E\n\r]/g, ' ').replace(/\s{3,}/g, '\n').trim();
    }
  } catch {
    // fall through
  }

  return '';
}

/**
 * Pattern-based resume parser. Extracts structured data from raw resume text.
 * Uses regex patterns to identify contact info, skills, experience, education.
 */
export function parseResumeText(rawText: string): ResumeData {
  const lines = rawText.split('\n').map((l) => l.trim()).filter(Boolean);
  const data: ResumeData = {
    contact: {},
    summary: '',
    skills: [],
    experience: [],
    education: [],
    certifications: [],
  };

  // Email
  const emailMatch = rawText.match(/[\w.+-]+@[\w-]+\.[\w.-]+/);
  if (emailMatch) data.contact!.email = emailMatch[0];

  // Phone
  const phoneMatch = rawText.match(/(\+?\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3,4}[-.\s]?\d{4}/);
  if (phoneMatch) data.contact!.phone = phoneMatch[0];

  // LinkedIn
  const linkedinMatch = rawText.match(/linkedin\.com\/(in\/[\w-]+)/i);
  if (linkedinMatch) data.contact!.linkedin = `linkedin.com/${linkedinMatch[1]}`;

  // Website
  const websiteMatch = rawText.match(/(https?:\/\/[\w.-]+\.[a-z]{2,}[^\s]*)/i);
  if (websiteMatch) data.contact!.website = websiteMatch[0];

  // Name — first non-empty line that's not an email/phone
  for (const line of lines.slice(0, 5)) {
    if (
      !line.includes('@') &&
      !line.match(/\d{3,}/) &&
      !line.match(/^(resume|cv|curriculum)/i) &&
      line.length > 3 &&
      line.length < 60 &&
      line.split(' ').length <= 4
    ) {
      data.contact!.name = line;
      break;
    }
  }

  // Skills section
  const skillsIdx = lines.findIndex((l) => /^(technical\s+)?skills?[:\s]/i.test(l));
  if (skillsIdx >= 0) {
    const skillsLine = lines[skillsIdx].replace(/^(technical\s+)?skills?[:\s]*/i, '');
    const afterLine = lines.slice(skillsIdx + 1).slice(0, 5).join(' ');
    const allSkills = (skillsLine + ' ' + afterLine)
      .split(/[,;|•·]\s*|\s{2,}/)
      .map((s) => s.trim())
      .filter((s) => s.length > 1 && s.length < 40);
    data.skills = [...new Set(allSkills)] as string[];
  }

  // If no skills section found, try to extract from common tech keywords
  if (!data.skills || data.skills.length === 0) {
    const techKeywords = [
      'JavaScript', 'TypeScript', 'Python', 'Java', 'C++', 'C#', 'React', 'Angular', 'Vue',
      'Node.js', 'Express', 'Django', 'Flask', 'FastAPI', 'Spring', 'SQL', 'PostgreSQL',
      'MySQL', 'MongoDB', 'Redis', 'Docker', 'Kubernetes', 'AWS', 'Azure', 'GCP',
      'Git', 'CI/CD', 'Jenkins', 'REST', 'GraphQL', 'HTML', 'CSS', 'Tailwind',
      'Machine Learning', 'TensorFlow', 'PyTorch', 'Pandas', 'NumPy', 'Scikit-learn',
      'Data Analysis', 'Tableau', 'Power BI', 'Excel', 'Leadership', 'Agile', 'Scrum',
      'Project Management', 'Communication', 'Teamwork', 'Problem Solving',
    ];
    const found = techKeywords.filter((kw) =>
      new RegExp(`\\b${kw.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`, 'i').test(rawText)
    );
    data.skills = found;
  }

  // Experience section
  const expIdx = lines.findIndex((l) => /^(work\s+)?experience[:\s]/i.test(l));
  if (expIdx >= 0) {
    let i = expIdx + 1;
    while (i < lines.length && !/^(education|certifications?|projects?|skills?)[:\s]/i.test(lines[i])) {
      const line = lines[i];
      if (line.length > 3) {
        const dateMatch = line.match(/(\d{4})\s*[-–]\s*(\d{4}|present|current)/i);
        if (dateMatch) {
          data.experience!.push({
            title: line.replace(/\d{4}.*$/, '').trim() || 'Position',
            start_date: dateMatch[1],
            end_date: dateMatch[2] || '',
            description: lines.slice(i + 1, i + 4).join(' ').slice(0, 300),
          });
          i += 4;
          continue;
        }
        if (!data.experience!.length || data.experience![data.experience!.length - 1].company) {
          data.experience!.push({ company: line, title: '', description: '' });
        } else {
          data.experience![data.experience!.length - 1].company = line;
        }
      }
      i++;
    }
  }

  // Education section
  const eduIdx = lines.findIndex((l) => /^education[:\s]/i.test(l));
  if (eduIdx >= 0) {
    let i = eduIdx + 1;
    while (i < lines.length && !/^(experience|certifications?|projects?|skills?)[:\s]/i.test(lines[i]) && i < eduIdx + 10) {
      const line = lines[i];
      if (line.length > 3) {
        const degreeMatch = line.match(/(b\.?sc\.?|b\.?tech\.?|m\.?sc\.?|m\.?tech\.?|mba|ph\.?d|bachelor|master|diploma)/i);
        data.education!.push({
          institution: line.replace(/(b\.?sc\.?|b\.?tech\.?|m\.?sc\.?|m\.?tech\.?|mba|ph\.?d|bachelor|master|diploma).*$/i, '').trim() || line,
          degree: degreeMatch ? degreeMatch[0] : '',
        });
      }
      i++;
    }
  }

  // Summary — first paragraph after name that's not contact info
  for (const line of lines) {
    if (
      line.length > 50 &&
      !line.includes('@') &&
      !line.match(/^\+?\d/) &&
      !line.match(/^(skills?|experience|education|certifications?)/i)
    ) {
      data.summary = line.slice(0, 500);
      break;
    }
  }

  return data;
}

export function extractSkillsFromData(data: ResumeData): string[] {
  const skills = new Set<string>(data.skills || []);
  if (data.experience) {
    for (const exp of data.experience) {
      if (exp.description) {
        const techKeywords = [
          'JavaScript', 'TypeScript', 'Python', 'Java', 'React', 'Node.js', 'SQL',
          'AWS', 'Docker', 'Kubernetes', 'Git', 'REST', 'GraphQL', 'HTML', 'CSS',
        ];
        for (const kw of techKeywords) {
          if (new RegExp(`\\b${kw}\\b`, 'i').test(exp.description)) skills.add(kw);
        }
      }
    }
  }
  return [...skills];
}

File: Synapse/backend/app/services/resume_parser.py
import re
from typing import Optional


def parse_resume_text(raw_text: str) -> dict:
    lines = [l.strip() for l in raw_text.split("\n") if l.strip()]
    data = {
        "contact": {},
        "summary": "",
        "skills": [],
        "experience": [],
        "education": [],
        "certifications": [],
    }

    email_match = re.search(r'[\w.+-]+@[\w-]+\.[\w.-]+', raw_text)
    if email_match:
        data["contact"]["email"] = email_match.group()

    phone_match = re.search(r'(\+?\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3,4}[-.\s]?\d{4}', raw_text)
    if phone_match:
        data["contact"]["phone"] = phone_match.group()

    linkedin_match = re.search(r'linkedin\.com/(in/[\w-]+)', raw_text, re.I)
    if linkedin_match:
        data["contact"]["linkedin"] = f"linkedin.com/{linkedin_match.group(1)}"

    website_match = re.search(r'(https?://[\w.-]+\.[a-z]{2,}[^\s]*)', raw_text, re.I)
    if website_match:
        data["contact"]["website"] = website_match.group()

    for line in lines[:5]:
        if (
            "@" not in line
            and not re.search(r'\d{3,}', line)
            and not re.match(r'^(resume|cv|curriculum)', line, re.I)
            and 3 < len(line) < 60
            and len(line.split()) <= 4
        ):
            data["contact"]["name"] = line
            break

    skills_idx = next(
        (i for i, l in enumerate(lines) if re.match(r'^(technical\s+)?skills?[:\s]', l, re.I)),
        -1,
    )
    if skills_idx >= 0:
        skills_line = re.sub(r'^(technical\s+)?skills?[:\s]*', '', lines[skills_idx], flags=re.I)
        after_lines = " ".join(lines[skills_idx + 1: skills_idx + 6])
        all_skills = re.split(r'[,;|•·]\s*|\s{2,}', skills_line + " " + after_lines)
        data["skills"] = list({s.strip() for s in all_skills if 1 < len(s.strip()) < 40})

    if not data["skills"]:
        tech_keywords = [
            'JavaScript', 'TypeScript', 'Python', 'Java', 'C++', 'C#', 'React', 'Angular', 'Vue',
            'Node.js', 'Express', 'Django', 'Flask', 'FastAPI', 'Spring', 'SQL', 'PostgreSQL',
            'MySQL', 'MongoDB', 'Redis', 'Docker', 'Kubernetes', 'AWS', 'Azure', 'GCP',
            'Git', 'CI/CD', 'Jenkins', 'REST', 'GraphQL', 'HTML', 'CSS', 'Tailwind',
            'Machine Learning', 'TensorFlow', 'PyTorch', 'Pandas', 'NumPy', 'Scikit-learn',
        ]
        data["skills"] = [
            kw for kw in tech_keywords
            if re.search(rf'\b{re.escape(kw)}\b', raw_text, re.I)
        ]

    exp_idx = next(
        (i for i, l in enumerate(lines) if re.match(r'^(work\s+)?experience[:\s]', l, re.I)),
        -1,
    )
    if exp_idx >= 0:
        i = exp_idx + 1
        while i < len(lines) and not re.match(r'^(education|certifications?|projects?|skills?)[:\s]', lines[i], re.I):
            line = lines[i]
            if len(line) > 3:
                date_match = re.search(r'(\d{4})\s*[-–]\s*(\d{4}|present|current)', line, re.I)
                if date_match:
                    data["experience"].append({
                        "title": re.sub(r'\d{4}.*$', '', line).strip() or "Position",
                        "start_date": date_match.group(1),
                        "end_date": date_match.group(2) or "",
                        "description": " ".join(lines[i + 1: i + 4])[:300],
                    })
                    i += 4
                    continue
                if not data["experience"] or data["experience"][-1].get("company"):
                    data["experience"].append({"company": line, "title": "", "description": ""})
                else:
                    data["experience"][-1]["company"] = line
            i += 1

    edu_idx = next(
        (i for i, l in enumerate(lines) if re.match(r'^education[:\s]', l, re.I)),
        -1,
    )
    if edu_idx >= 0:
        i = edu_idx + 1
        while i < len(lines) and not re.match(r'^(experience|certifications?|projects?|skills?)[:\s]', lines[i], re.I) and i < edu_idx + 10:
            line = lines[i]
            if len(line) > 3:
                degree_match = re.search(
                    r'(b\.?sc\.?|b\.?tech\.?|m\.?sc\.?|m\.?tech\.?|mba|ph\.?d|bachelor|master|diploma)',
                    line, re.I,
                )
                data["education"].append({
                    "institution": re.sub(
                        r'(b\.?sc\.?|b\.?tech\.?|m\.?sc\.?|m\.?tech\.?|mba|ph\.?d|bachelor|master|diploma).*$',
                        '', line, flags=re.I,
                    ).strip() or line,
                    "degree": degree_match.group() if degree_match else "",
                })
            i += 1

    for line in lines:
        if (
            len(line) > 50
            and "@" not in line
            and not re.match(r'^\+?\d', line)
            and not re.match(r'^(skills?|experience|education|certifications?)', line, re.I)
        ):
            data["summary"] = line[:500]
            break

    return data


def extract_skills_from_data(data: dict) -> list[str]:
    skills = set(data.get("skills", []))
    for exp in data.get("experience", []):
        desc = exp.get("description", "")
        tech_keywords = [
            'JavaScript', 'TypeScript', 'Python', 'Java', 'React', 'Node.js', 'SQL',
            'AWS', 'Docker', 'Kubernetes', 'Git', 'REST', 'GraphQL', 'HTML', 'CSS',
        ]
        for kw in tech_keywords:
            if re.search(rf'\b{kw}\b', desc, re.I):
                skills.add(kw)
    return list(skills)
